### ISSUE TITLE
Multiple build failures

### ISSUE THREAD

**Issue by arkonbob** [4ch1m/ansible-betterbird#2]
Hi, 

Thanks for putting this together, I'm keen to try it out. I've hit a few stumbling blocks so far:

1. The first time I ran the build script, it wasn't able to install the packages. I'm on OpenSuSE Tumbleweed (details below). I had to rename a couple of the packages in `source/python/mozboot/mozboot/opensuse.py` to match those in the OpenSuSE repos: npm -> npm16, nodejs -> nodejs16
2. The next run timed out waiting for my sudo password at the bootstrap stage (I had mistakenly thought I could fire-and-forget). I guess this is just sudo / Ansible behaviour and there is nothing to be done...
3. Re-running the build, it failed repeating the application of patches (I assume due to (2).
3. I removed the already completed steps prior to bootstrap from betterbird.yml and re-ran. The script failed silently.
4. I ran the bootstrap stage manually:
```./mach --no-interactive -vvv bootstrap --application-choice "Firefox for Desktop"```

This failed with the following output:
```
You can only create a mach environment with the system Python. Re-run this `mach` command with the system Python.
Error running mach:

    ['--no-interactive', '-vvv', 'bootstrap', '--application-choice', 'Firefox for Desktop']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.
You can invoke |./mach busted| to check if this issue is already on file. If it
isn't, please use |./mach busted file bootstrap| to report it. If |./mach busted| is
misbehaving, you can also inspect the dependencies of bug 1543241.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

subprocess.CalledProcessError: Command '['/home/rob/.mozbuild/_virtualenvs/mach/bin/python', '/home/rob/dev/ansible-betterbird/source/mach', 'create-mach-environment']' returned non-zero exit status 1.

  File "/home/rob/dev/ansible-betterbird/source/python/mozboot/mozboot/mach_commands.py", line 55, in bootstrap
    bootstrapper.bootstrap(self.settings)
```

I'm at a loss as to where to go from here, could you suggest how to proceed?

My system:
  LSB Version:    n/a
  Distributor ID: openSUSE
  Description:    openSUSE Tumbleweed
  Release:        20210912
  Codename:       n/a

**Comment by 4ch1m** [comment]
> The first time I ran the build script, it wasn't able to install the packages. I'm on OpenSuSE Tumbleweed (details below). I had to rename a couple of the packages in source/python/mozboot/mozboot/opensuse.py to match those in the OpenSuSE repos: npm -> npm16, nodejs -> nodejs16

This is out of scope of this Ansible playbook.
Should be addressed to the core Thunderbird maintainers/developers.

>  The next run timed out waiting for my sudo password at the bootstrap stage (I had mistakenly thought I could fire-and-forget). I guess this is just sudo / Ansible behaviour and there is nothing to be done...

All right.
Maybe we should add "become: yes" to the bootstrap-task as well.
I'll try to figure this out.

>     Re-running the build, it failed repeating the application of patches (I assume due to (2).
>     I removed the already completed steps prior to bootstrap from betterbird.yml and re-ran. The script failed silently.
>     I ran the bootstrap stage manually:
>     ./mach --no-interactive -vvv bootstrap --application-choice "Firefox for Desktop"

What exactly are you doing when you're "Re-running the build"?
You should always re-run the whole Ansible playbook... not just manually execute certain steps yourself on a console.
The playbook will always try to reset everything and start fresh. Making sure there are no "left overs" that could break the workflow.

**Comment by arkonbob** [comment]
I meant I was re-running the script with `./betterbird.sh`, which did not seem to be starting afresh. I have restored `betterbird.yml` to its original state and re-run `betterbird.sh` to reproduce the error I described in (3):
```
TASK [apply patches | mozilla - apply all patch files] **************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "hg qpush -a", "delta": "0:00:04.069062", "end": "2021-09-15 10:22:48.005810", "msg": "non-zero return code", "rc": 255, "start": "2021-09-15 10:22:43.936748", "stderr": "cleaning up working directory...\ndone\nabort: qpush exactly duplicates child changeset", "stderr_lines": ["cleaning up working directory...", "done", "abort: qpush exactly duplicates child changeset"], "stdout": "(working directory not at a head)\napplying 08-branding-m-c.patch\nreverting ../../toolkit/mozapps/installer/windows/nsis/makensis.mk", "stdout_lines": ["(working directory not at a head)", "applying 08-branding-m-c.patch", "reverting ../../toolkit/mozapps/installer/windows/nsis/makensis.mk"]}

PLAY RECAP **********************************************************************************************************************************************************************************
localhost                  : ok=26   changed=12   unreachable=0    failed=1    skipped=3    rescued=0    ignored=0   
```

**Comment by 4ch1m** [comment]
Are you sure you're on the latest commit?
This problem should be fixed with the recent updates.

**Comment by arkonbob** [comment]
That was using a fresh checkout this morning.

```
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
$ git log | head                                                            
commit 00820fd1ecc916d3fa391a6471f98839dae94bfc
Author: Achim <git@achimonline.de>
Date:   Sat Sep 11 16:11:19 2021 +0200

    fix "when"-condition

commit 53e7304ba879a6342f747b5ea6ccce8ac5e3fe53
Author: Achim <git@achimonline.de>
Date:   Sat Sep 11 15:25:07 2021 +0200
```

**Comment by 4ch1m** [comment]
Ok.
Not sure if I can assist you with this issue though.
It always works for me now.
Others also reported successful builds with the Ansible playbook.
But let me know if you found the issue.
Thanks.

**Comment by arkonbob** [comment]
A couple of days later I tried again and got it working. Many thanks for packaging this up for the community, it's great to see things in Thunderbird getting fixed! Yesterday I decided to update to the latest and thought I would feedback here for anyone else wanting to build on OpenSuSE. My notes:

The following patch needs to be applied if you have Wayland header files present:
https://hg.mozilla.org/integration/autoland/rev/f7e2cf808d30 
Add the following at the foot of the file `./source/widget/gtk/mozwayland/mozwayland.c`:
```
MOZ_EXPORT struct wl_proxy* wl_proxy_marshal_flags(                                                                        struct wl_proxy* proxy, uint32_t opcode,
    const struct wl_interface* interface, uint32_t version, uint32_t flags,
    ...) {
    return NULL;
}
```
To enable ansible build logging to file, create ansible.cfg in the project directory:
```
[defaults]
log_path=./build.log
```
Run ./betterbird.sh 
After the first failure, edit `source/python/mozboot/mozboot/opensuse.py` and add '16' to the node and npm package names:
```
...
SYSTEM_PACKAGES = [
        "nodejs16",
        "npm16",
        "which",
        "rpmconf",
        "libcurl-devel",
        "libpulse-devel",
    ]
...
```
Edit betterbird.yml and comment out all the tasks before `- name: bootstrap`.
Re-run ./betterbird.sh

If there are issues with the rust environment, run:
```
rustup install stable
rustup default stable
```
